src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -109,34 +104,48 @@ public static TensorFlowTransform Create(IHostEnvironment env, ModelLoadContext
             Contracts.CheckValue(env, nameof(env));
             env.CheckValue(ctx, nameof(ctx));
             ctx.CheckAtModel(GetVersionInfo());
+
             // *** Binary format ***
+            // int: indicator for frozen models
             // stream: tensorFlow model.
             // int: number of input columns
             // for each input column
             //   int: id of int column name
             // int: number of output columns
             // for each output column
             //   int: id of output column name
-            byte[] modelBytes = null;
-            if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
-                throw env.ExceptDecode();
-            var numInputs = ctx.Reader.ReadInt32();
-            env.CheckDecode(numInputs > 0);
-            string[] inputs = new string[numInputs];
-            for (int j = 0; j < inputs.Length; j++)
-                inputs[j] = ctx.LoadNonEmptyString();
+            var isFrozen = ctx.Reader.ReadInt32();
+            if (isFrozen == 1)
+            {
+                byte[] modelBytes = null;
+                if (!ctx.TryLoadBinaryStream("TFModel", r => modelBytes = r.ReadByteArray()))
+                    throw env.ExceptDecode();
 
-            bool isMultiOutput = ctx.Header.ModelVerReadable >= 0x00010002;
-            var numOutputs = 1;
-            if (isMultiOutput)
-                numOutputs = ctx.Reader.ReadInt32();
+                var io = ModelInputsOutputs(env, ctx);
+                return new TensorFlowTransform(env, modelBytes, io.Item1, io.Item2, (isFrozen == 1));
+            }
+            else
+            {
```

Comment:
> { [](start = 11, length = 2)

[observation] The Train() method of the LearningPipeline API ends up calling Save() and Create() multiple times.

For un-frozen models we need a particular directory structure so we can create the TFSession correctly using the TFSession.FromSavedModel API  (line 215)

In this call to Create() we do the following :
(a) load the binary stream
(b) write the binary stream out to a  *zip file*
(c) extract the zip file to a *temporary folder* TempSavedModelDirName  (_savedModel)
(d)  create the session  
(e)  delete the *zip file*

We cannot delete the *temporary folder* because in a subsequent call to Save()  we are  writing out the contents of the *temporary folder*  as a byte stream.

So for this scenario, what we have currently is that we *define* a unique location for the *temporary folder* -- and it gets cleaned up only when required, inside the Create() call. Line 139.  

@ericstj

Reply:
You should avoid the temporary zip file.  If you can get a seekable stream from ctx, just pass that into the ZipArchive constructor: [https://msdn.microsoft.com/es-es/library/system.io.compression.ziparchive.ziparchive(v=vs.110).aspx](https://msdn.microsoft.com/es-es/library/system.io.compression.ziparchive.ziparchive(v=vs.110).aspx).  It looks like you can do this by accessing the BaseStream  on the binary reader passed into the TryLoadBinaryStream action.

You should also avoid allocating an extra byte array for the zip and instead do all the reading within the callback from TryLoadBinaryStream, since it looks like TryLoadBinaryStream will close the stream: [https://github.com/dotnet/machinelearning/blob/0d5e317b2196812edf976cee51d2ed6534c45ad0/src/Microsoft.ML.Data/Model/ModelLoading.cs#L292](https://github.com/dotnet/machinelearning/blob/0d5e317b2196812edf976cee51d2ed6534c45ad0/src/Microsoft.ML.Data/Model/ModelLoading.cs#L292).  So within this callback, new up the archive, then call archive.ExtractToDirectory extension method: [https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfileextensions.extracttodirectory?view=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfileextensions.extracttodirectory?view=netframework-4.7.2).

I think you still need to have a temporary directory for the model files themselves.  It looks like this is a requirement of TensorFlow [https://github.com/tensorflow/tensorflow/blob/3be04971716fcaf0c11ad9262e60efa428553e14/tensorflow/c/c_api.h#L1367](https://github.com/tensorflow/tensorflow/blob/3be04971716fcaf0c11ad9262e60efa428553e14/tensorflow/c/c_api.h#L1367).  It requires the model to be in a directory and doesn't provide a mechanism for loading from some sort of stream/memory representation.  So you'll still need a temp path that you'll have to clean up.

Ideally you should provide a way for this path to be specified, rather than always call GetTempPath behind the scenes.  I think ML.NET has a pattern for this, if not, perhaps consider a constructor overload that allows for specification of a temp path root.

Typically with things like this you should treat the temporary folder as state managed by the object and then cleaned up in the dispose/finalizer.



---
In reply to: [217123233](https://github.com/dotnet/machinelearning/pull/853#discussion_r217123233) [](ancestors = 217123233)

Reply:
I tried both of the suggestions above, but they seem to have some limitations with respect to this transform. I have since modified the logic of load&restore to do things in-memory.

Am documenting the issues hit with the above suggestion:

(1) the ZipArchive approach looked promising to avoid the temporary zip file. However it works on the entire stream, rather than on a subset of the stream.
(2) currently the TFTransform does not dispose the TFSession object.  So attempts to cleanup the temp folder fail with 'file still in use' errors. The approach of using unique location for the temporary folder is problematic in general (e.g. during CV or if pipeline contains multiple TFTransforms)

I have modified the logic so we do the ML.NET serialization/de-serialization using in-memory data structures.
  

---
In reply to: [217145506](https://github.com/dotnet/machinelearning/pull/853#discussion_r217145506) [](ancestors = 217145506,217123233)

Reply:
You should file an issue on #2.  

I think this approach looks much better than zip-in-a-zip approach before.  One change I'd suggest is that instead of reading to byte array and storing in memory, you  create the file within the callback (var fs = new FileStream(...)) and use br.BaseStream.CopyTo(fs, fileLength);

One thing I notice that could be problematic for the frozen case above (not related to this change) is that it seems to use Graph.Import to import a entire model as a managed byte array.  I can imagine that this will potentially hit the upper limit of managed array sizes for large models.  I see that TF's graph import functionality expects a single buffer ([https://github.com/tensorflow/tensorflow/blob/3be04971716fcaf0c11ad9262e60efa428553e14/tensorflow/c/c_api.h#L1018-L1020](https://github.com/tensorflow/tensorflow/blob/3be04971716fcaf0c11ad9262e60efa428553e14/tensorflow/c/c_api.h#L1018-L1020)).

---
In reply to: [217447210](https://github.com/dotnet/machinelearning/pull/853#discussion_r217447210) [](ancestors = 217447210,217145506,217123233)

Reply:
The main reason for reading to byte array and storing in memory is  to handle the scenario  where  we saw the sequence of  Save1() -> Create1()  -> Save2() -> Create2() being called

So when Create1() is called, we load up the contents into a dictionary.  So when Save2()  is called we use the dictionary to write out the contents.  

Because of issue #906 (not being able to clean-up unmanaged resources ) I am not sure we can do away with the in-memory approach. Thoughts ?

---
In reply to: [217462971](https://github.com/dotnet/machinelearning/pull/853#discussion_r217462971) [](ancestors = 217462971,217447210,217145506,217123233)

Reply:
adopted the stream copy based approach instead of reading to byte array 

also added a  finalizer that closes the session (if it isn’t closed) and deletes the temporary directory

---
In reply to: [217545311](https://github.com/dotnet/machinelearning/pull/853#discussion_r217545311) [](ancestors = 217545311,217462971,217447210,217145506,217123233)

Reply:
@eerhardt , @danmosemsft , would this be a security risk? To load the TensorFlow model here, we create a temporary directory and copy some files to it, and then load the model from that directory.

Reply:
followed the recommendation of the .NET security team to address this risk

---
In reply to: [218159541](https://github.com/dotnet/machinelearning/pull/853#discussion_r218159541) [](ancestors = 218159541)